src/interface.jl

```diff
@@ -31,111 +46,201 @@ ADBackend(::Val) = error("The requested AD backend is not available. Make sure t
 ######################
 # Bijector interface #
 ######################
-"Abstract type for a bijector."
-abstract type AbstractBijector end
+"""
+
+Abstract type for a transformation.
+
+# Implementing
+
+A subtype of `Transform` of should at least implement [`transform(b, x)`](@ref).
+
+If the `Transform` is also invertible:
+- Required:
+  - [`invertible`](@ref): should return instance of [`Invertible`](@ref) or [`NotInvertible`](@ref).
```

Comment:
Why not just use `InverseFunctions`? If `inverse` is not defined, it returns by default `NoInverse` (IIRC, name might be slightly different), so you can check if the function is invertible (or rather, if `ìnverse` is defined).

Reply:
Ah, happy to! Just wasn't aware it did this. Will change accordingly.

Reply:
Okay so one thing I find non-ideal with `InverseFunctions.NoInverse` is that `inverse` sometimes performs operations, e.g. if you didn't implement the inverse of `Scale` lazily, you might have to invert a matrix just to check if it's invertible or not. I will make the change though, but this is annoying :confused:

Reply:
In addition, the behavior I _expect_ when calling `inverse(f)` is not to get some lazy expression. I expect to get a callable back that computes `inverse(f)` in the most efficient way.

For example,

``` julia
inverse(Scale(A))
```

I'd expect to give me a `Scale(inv(A))`, not `Inverse{<:Scale}` which uses `A \ x` under the hood.

I think to support the lazy versions, we should instead add methods such as `transform_inverse` and `with_logabsdet_jacobian_inverse`.

Reply:
> In addition, the behavior I expect when calling inverse(f) is not to get some lazy expression. I expect to get a callable back that computes inverse(f) in the most efficient way.

Isn't that was `inverse` is supposed to return? For instance, `inverse(exp) = log` - you get a callable back that computes the inverse of `exp`?

> I'd expect to give me a Scale(inv(A)), not Inverse{<:Scale} which uses A \ x under the hood.

If you want `Scale(inv(A))`, you can just define `inverse(s::Scale) = Scale(inv(s.A))`, no? And if you want `Inverse{<:Scale}`, you would define `inverse(s::Scale) = Inverse(s)` and `(t::Inverse{<:Scale})(x) = t.x.A \ x`?